aws-s3-transfer-manager/src/operation/upload_objects/worker.rs

```diff
@@ -51,8 +51,23 @@ pub(super) async fn list_directory_contents(
     while let Some(entry) = walker.next().await {
         let job = match entry {
             Ok(entry) => {
-                let metadata = tokio::fs::metadata(entry.path()).await?;
-                if !(filter.predicate)(&UploadFilterItem::new(entry.path(), metadata.clone())) {
+                let symlink_metadata = tokio::fs::symlink_metadata(entry.path()).await?;
+                let metadata = if symlink_metadata.is_symlink() {
+                    if input.follow_symlinks {
+                        tokio::fs::metadata(entry.path()).await?
+                    } else {
+                        continue;
+                    }
+                } else {
+                    // In this branch, we know `symlink_metadata` does not represent a symlink link,
+                    // so it can return true for either `is_dir()` or `is_file()`.
+                    symlink_metadata
+                };
+                let filter_item = UploadFilterItem::builder()
+                    .path(entry.path())
+                    .metadata(metadata.clone())
+                    .build();
+                if !(filter.predicate)(&filter_item) {
```

Comment:
Are we doing anything different in this loop based on whether `entry` is a file or directory? If so, I don't see where

Reply:
No we're not. Both types, a directory and a file, are presented to the filter.

Reply:
No we're not. Both types, a directory and a file, are yielded in the loop. But the default filter only includes files.